src/test/rgw/test_rgw_iam_policy.cc

```diff
@@ -1311,3 +1312,537 @@ TEST(set_cont_bits, iamconsts)
+
+  ASSERT_NO_THROW(p = Policy(cct.get(), arbitrary_tenant,
+			     bufferlist::static_from_string(delimiter_allow_policy)));
+  ASSERT_TRUE(p);
+
+  EXPECT_EQ(p->text, delimiter_allow_policy);
+  EXPECT_EQ(p->version, Version::v2012_10_17);
+  EXPECT_EQ(p->statements[0].effect, Effect::Allow);
+  EXPECT_EQ(p->statements[0].resource.begin()->partition, Partition::aws);
+  Action_t act;
+  act[s3ListBucket] = 1;
+  EXPECT_EQ(p->statements[0].action, act);
+  ASSERT_FALSE(p->statements[0].conditions.empty());
+  EXPECT_EQ(p->statements[0].conditions[0].vals[0], "MP");
+}
+
+TEST_F(PolicyConditionTest, EvalDelimiter) {
+  auto allowp = Policy(cct.get(), arbitrary_tenant,
+		   bufferlist::static_from_string(delimiter_allow_policy));
+  auto denyp = Policy(cct.get(), arbitrary_tenant,
+		   bufferlist::static_from_string(delimiter_deny_policy));
+  Environment e;
+  e.emplace("s3:delimiter", "MP");
+
+  EXPECT_EQ(allowp.eval_conditions(e),
+	    Effect::Allow);
+  // Evaluate condition is true or false. Checks the condition element
+  EXPECT_EQ(denyp.eval_conditions(e),
+	    Effect::Deny);
+}
+
+TEST_F(PolicyConditionTest, ParsePrefix) {
+  boost::optional<Policy> p;
+
+  ASSERT_NO_THROW(p = Policy(cct.get(), arbitrary_tenant,
+			     bufferlist::static_from_string(prefix_allow_policy)));
+  ASSERT_TRUE(p);
+
+  EXPECT_EQ(p->text, prefix_allow_policy);
+  EXPECT_EQ(p->version, Version::v2012_10_17);
+  EXPECT_EQ(p->statements[0].effect, Effect::Allow);
+  EXPECT_EQ(p->statements[0].resource.begin()->partition, Partition::aws);
+  Action_t act;
+  act[s3ListBucket] = 1;
+  EXPECT_EQ(p->statements[0].action, act);
+  ASSERT_FALSE(p->statements[0].conditions.empty());
+  EXPECT_EQ(p->statements[0].conditions[0].vals[0], "");
+  EXPECT_EQ(p->statements[0].conditions[0].vals[1], "home/");
+}
+
+TEST_F(PolicyConditionTest, EvalPrefix) {
+  auto allowp = Policy(cct.get(), arbitrary_tenant,
+		   bufferlist::static_from_string(prefix_allow_policy));
+  auto denyp = Policy(cct.get(), arbitrary_tenant,
+		   bufferlist::static_from_string(prefix_deny_policy));
+  Environment e;
+  e.emplace("s3:prefix", "home/");
+
+  EXPECT_EQ(allowp.eval_conditions(e),
+	    Effect::Allow);
+  // Evaluate condition is true or false. Checks the condition element
+  EXPECT_EQ(denyp.eval_conditions(e),
+	    Effect::Deny);
+}
+string PolicyConditionTest::username_allow_policy = R"(
+{
+  "Version": "2012-10-17",
+  "Statement": {
+    "Effect": "Allow",
+    "Action": "s3:ListBucket",
+    "Resource": "arn:aws:s3:::testbkt",
+     "Condition" : { "StringEquals" : { "aws:username" : "policy_user" }}
+  }
+}
+)";
+string PolicyConditionTest::username_deny_policy = R"(
+{
+  "Version": "2012-10-17",
+  "Statement": {
+    "Effect": "Deny",
+    "Action": "s3:ListBucket",
+    "Resource": "arn:aws:s3:::testbkt",
+    // Evaluate condition is true or false. Checks the condition element
+     "Condition" : { "StringEquals" : { "aws:username" : "policy_user1" }}
```

Comment:
who is creating testbkt. if some other user creating testbkt, policy_user1 already have no access to this bucket,
Are we taking care of that?

Reply:
Here no bucket is getting created, just the condition is getting evaluated, just calling the function directly , no RGW is running